bftengine/include/bftengine/ReservedPages.hpp

```diff
@@ -0,0 +1,111 @@
+// Concord
+//
+// Copyright (c) 2018 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <cstdint>
+#include <string>
+#include <typeindex>
+#include <algorithm>
+#include <tuple>
+#include <vector>
+#include <iostream>
+#include "demangle.hpp"
+#include "assertUtils.hpp"
+
+namespace bftEngine {
+
+class ReservedPages {
+ public:
+  static uint32_t totalNumberOfPages() {
+    uint32_t numPages{0};
+    for (auto& elt : registry()) numPages += std::get<2>(elt);
+
+    return numPages;
+  }
+
+ protected:
+  // [idx, typeid, numberOfPages]
+  typedef std::vector<std::tuple<std::uint8_t, std::type_index, std::uint32_t>> Registry;
+
+  static Registry& registry() {
+    static Registry registry_;
+    return registry_;
+  }
+};
+/**
+ * Facility for registering Reserved Pages clients
+ *
+ * Become a Reserved Pages client by sub-classing:
+ * class MyClass: public ResPagesClient<MyClass, idx, requiredNumberOfPages> {};
+ * where idx is a unique integer used for dividing a reserved pages space into sub-spaces.
```

Comment:
No, but he/she can search for it and also there will be an immediate runtime error upon startup if there's a collision.
I haven't found a more elegant way of doing it while enforcing the namespace order.

Reply:
I think the problem is that you created an open type (via the use of inheritance), but all the users of the of reserved pages are known at startup time, and need to choose an index a priori. The latter implies that this is in fact a closed type and a different mechanism should be used, like an enum or variant.

It seems somewhat wrong to have users search for/learn of in-use names during startup, even though they have to pick a value for an index. Either the users should pick, in which case there should be a central registry in one place, where all users are known, or the registry is truly open and dynamic, and users register their clients and get back an index. 

My preference would be to remove the inheritance and the need for worrying about typeids and just create a central registry that allocates the vector at startup. This doesn't mean that how reserved pages are used is known by the registry, just what components are using them and their order in the registry vector. Each component can still register itself before reserved pages are used, and can still access it's place in the registry to set its number of pages without taking a lock. This should work fine, since only concord-bft components directly access and manipulate reserved pages, and so we can list them all in ReservedPages.hpp.

The big advantage to the central registry/enum is that we won't have to try to remember which components use reserved pages. We can just look in the registry header. It will also greatly simplify this code. Clients now only need to take their index from the registry and call `setNumResPages`.  You can even have a safety check for when you start the code that each component has indeed set its reserved pages, since the registry knows which components should exist.

If there's a  reason why the registry must allow for arbitrary reserved pages clients, or I'm missing something else, please let me know.